velox/functions/lib/aggregates/MinMaxAggregateBase.cpp

```diff
@@ -0,0 +1,649 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/lib/aggregates/MinMaxAggregateBase.h"
+
+#include <limits>
+#include "velox/exec/AggregationHook.h"
+#include "velox/functions/lib/CheckNestedNulls.h"
+#include "velox/functions/lib/aggregates/Compare.h"
+#include "velox/functions/lib/aggregates/SimpleNumericAggregate.h"
+#include "velox/functions/lib/aggregates/SingleValueAccumulator.h"
+#include "velox/type/FloatingPointUtil.h"
+
+namespace facebook::velox::functions::aggregate {
+
+namespace {
+
+template <typename T>
+struct MinMaxTrait : public std::numeric_limits<T> {};
+
+template <typename T>
+class MinMaxAggregate : public SimpleNumericAggregate<T, T, T> {
+  using BaseAggregate = SimpleNumericAggregate<T, T, T>;
+
+ public:
+  explicit MinMaxAggregate(TypePtr resultType, TimestampPrecision precision)
+      : BaseAggregate(resultType), timestampPrecision_(precision) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(T);
+  }
+
+  int32_t accumulatorAlignmentSize() const override {
+    if constexpr (std::is_same_v<T, int128_t>) {
+      // Override 'accumulatorAlignmentSize' for UnscaledLongDecimal values as
+      // it uses int128_t type. Some CPUs don't support misaligned access to
+      // int128_t type.
+      return static_cast<int32_t>(sizeof(int128_t));
+    } else {
+      return 1;
+    }
+  }
+
+  bool supportsToIntermediate() const override {
+    return true;
+  }
+
+  void toIntermediate(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      VectorPtr& result) const override {
+    this->singleInputAsIntermediate(rows, args, result);
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    if constexpr (std::is_same_v<T, Timestamp>) {
+      // Truncate timestamps to corresponding precision.
+      BaseAggregate::template doExtractValues<
+          Timestamp>(groups, numGroups, result, [&](char* group) {
+        auto ts = *BaseAggregate::Aggregate::template value<Timestamp>(group);
+        if (timestampPrecision_ == TimestampPrecision::kMilliseconds) {
```

Comment:
nit: perhaps, move this logic into Timestamp class.

```
Timestamp::truncate(precision)
```

Reply:
Updated.